Lib/contextlib.py

```diff
@@ -197,6 +216,7 @@ def __exit__(self, typ, value, traceback):
             try:
                 raise RuntimeError("generator didn't stop after throw()")
             finally:
+                del self.exc_context
                 self.gen.close()
```

Comment:
Should we *always* clear self.exc_context?

```suggestion
            finally:
                del self.exc_context
            try:
                raise RuntimeError("generator didn't stop after throw()")
            finally:
                self.gen.close()
```

or, better, do that at the top of the method:

```py
    def __exit__(self, typ, value, traceback):
        exc_context = self.exc_context
        del self.exc_context
```

Reply:
Also, what happens if someone calls `__exit__` without a matching `__enter__` call first?  (Users do crazy stuff sometimes.)  Assuming that doesn't break for other reasons, won't uses of `self.exc_context` (incl. `del ...`) in this method break (raise AttributeError)?

Adjusting that last bit of suggested code:

```py
    def __exit__(self, typ, value, traceback):
        try:
            exc_context = self.exc_context
            del self.exc_context
        except AttributeError:
            exc_context = None
```

Reply:
If we always clear `self.exc_context` by setting it back to `None`, instead of deleting the attribute, the suggestion becomes:

```suggestion
            finally:
                self.exc_context = None
            try:
                raise RuntimeError("generator didn't stop after throw()")
            finally:
                self.gen.close()
```

or, better:

```py
    def __exit__(self, typ, value, traceback):
        exc_context = self.exc_context
        self.exc_context = None
```

Reply:
I was doing this to ensure that the class wasn't hanging on to the old exception and keeping objects alive unnecessarily. You're right though that I should just be dropping the reference instead of using `del`. I've applied your suggestion of setting a local variable to the exception, setting the instance variable to `None`, and letting the local variable naturally drop out of scope.

> Also, what happens if someone calls __exit__ without a matching __enter__ call first?

`self.exc_context` is set to `None` in the `_GeneratorContextManagerBase`'s `__init__` so it shouldn't be possible for an `AttributeError` to occur, even if `__exit__` is called before `__enter__`.